app/controllers/shopify_app/callback_controller.rb

```diff
@@ -55,24 +60,23 @@ def set_shopify_session
         token: token,
         api_version: ShopifyApp.configuration.api_version
       )
-      session[:shopify] = ShopifyApp::SessionRepository.store(session_store, user: associated_user)
-      session[:shopify_domain] = shop_name
-      session[:shopify_user] = associated_user
 
-      if ShopifyApp.configuration.per_user_tokens?
-        # Adds the user_session to the session to determine if the logged in user has changed
-        user_session = auth_hash&.extra&.session
-        raise IndexError, "Missing user session signature" if user_session.nil?
-        session[:user_session] = user_session
+      session[:shopify_user] = associated_user
+      if session[:shopify_user].present?
+        session[:user_id] = ShopifyApp::SessionRepository.store_user_session(session_store, associated_user)
+      else
+        session[:shop_id] = ShopifyApp::SessionRepository.store_shop_session(session_store)
       end
+      session[:shopify_domain] = shop_name
+      session[:user_session] = auth_hash&.extra&.session
     end
 
     def install_webhooks
       return unless ShopifyApp.configuration.has_webhooks?
 
       WebhooksManager.queue(
         shop_name,
-        token,
+        shop_session&.token || user_session.token,
```

Comment:
Should we use the safe navigator for the user token as well? I believe the session can be `nil` too.

Reply:
We should never be in a situation where we enqueue the job with a `nil` token as it will definitely break the job. I would want that to raise here before enqueing the job.